go/vt/vtctl/grpcvtctldserver/server.go

```diff
@@ -2835,6 +2856,230 @@ func (s *VtctldServer) ValidateKeyspace(ctx context.Context, req *vtctldatapb.Va
+
+			for _, alias := range aliases {
+				// Don't diff schemas for self
+				if referenceAlias == alias {
+					continue
+				}
+				aliasWg.Add(1)
+				go func(alias *topodatapb.TabletAlias) {
+					defer aliasWg.Done()
+					replicaSchema, err := schematools.GetSchema(ctx, s.ts, s.tmc, alias, nil, req.ExcludeTables, req.IncludeViews)
+					if err != nil {
+						aliasErrs.RecordError(fmt.Errorf("GetSchema(%v, nil, %v, %v) failed: %v", alias, req.ExcludeTables, req.IncludeViews, err))
+						return
+					}
+
+					tmutils.DiffSchema(topoproto.TabletAliasString(referenceAlias), referenceSchema, topoproto.TabletAliasString(alias), replicaSchema, &aliasErrs)
+				}(alias)
+			}
+			aliasWg.Wait()
+
+			if aliasErrs.HasErrors() {
+				for _, err := range aliasErrs.Errors {
+					errMessage := err.Error()
+					shardResp.Results = append(shardResp.Results, errMessage)
+					resp.Results = append(resp.Results, errMessage)
+				}
+			}
+
+			resp.ResultsByShard[shard] = &shardResp
+		}(shard)
+	}
+
+	wg.Wait()
+
+	return &resp, nil
+}
+
+// ValidateVersionKeyspace validates all versions are the same in all
+// tablets in a keyspace
+func (s *VtctldServer) ValidateVersionKeyspace(ctx context.Context, req *vtctldatapb.ValidateVersionKeyspaceRequest) (*vtctldatapb.ValidateVersionKeyspaceResponse, error) {
+	keyspace := req.Keyspace
+	// find all the shards
+	shards, err := s.ts.GetShardNames(ctx, keyspace)
+	resp := vtctldatapb.ValidateVersionKeyspaceResponse{
+		Results:        []string{},
+		ResultsByShard: make(map[string]*vtctldatapb.ValidateShardResponse, len(shards)),
+	}
+
+	if err != nil {
+		resp.Results = append(resp.Results, fmt.Sprintf("TopologyServer.GetShardNames(%v) failed: %v", keyspace, err))
+		return &resp, nil
+	}
+
+	// corner cases
+	if len(shards) == 0 {
+		resp.Results = append(resp.Results, fmt.Sprintf("no shards in keyspace %v", keyspace))
+		return &resp, nil
+	}
+
+	// find the reference version using the first shard's primary
+	si, err := s.ts.GetShard(ctx, keyspace, shards[0])
+	if err != nil {
+		resp.Results = append(resp.Results, fmt.Sprintf("unable to find primary shard %v/%v", keyspace, shards[0]))
+		return &resp, nil
+	}
+	if !si.HasPrimary() {
+		resp.Results = append(resp.Results, fmt.Sprintf("no primary in shard %v/%v", keyspace, shards[0]))
+		return &resp, nil
+	}
+
+	referenceAlias := si.PrimaryAlias
+	referenceVersion, err := s.GetVersion(ctx, &vtctldatapb.GetVersionRequest{TabletAlias: referenceAlias})
+	if err != nil {
+		resp.Results = append(resp.Results, fmt.Sprintf("unable to get reference version of first shard's primary tablet: %v", err))
+		return &resp, nil
+	}
+
+	// Mutex for ValidateVersionKeyspaceResponse
+	var m sync.Mutex
+
+	for _, shard := range shards {
+		m.Lock()
+		shardResp := vtctldatapb.ValidateShardResponse{
+			Results: []string{},
+		}
+
+		var (
+			// Mutex for ValidateShardResponse
+			sm sync.Mutex
+			// WaitGroup for tablets
+			wg sync.WaitGroup
```

Comment:
can we rename these variables to make the comments unnecessary? i feel like that will make the code further down easier to read as well

Reply:
yep! 😄 I totally agree